{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "R environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(dplyr)\n",
    "library(future)\n",
    "library(data.table)\n",
    "set.seed(1)\n",
    "plan(\"multiprocess\", workers = 8)\n",
    "options(future.globals.maxSize = 1000 * 1024^5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gbm <- readRDS('data/mapped_Ruiz2021_to_core_GBmap.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# writing tsv file for inferCNV  \n",
    "write.table(as.matrix(GetAssayData(object = gbm, slot = \"counts\")), \n",
    "            'data/gbm-count-matrix.tsv', sep = '\\t', row.names = T, \n",
    "            col.names = T, quote = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(gbm) <- gbm@meta.data$predicted.high_hierarchy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_file <- setDT(as.data.frame(Idents(gbm)), keep.rownames = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(anno_file, 'data/gbm-anno-file.tsv', sep = '\\t', col.names = F, row.names = F, quote = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### inferCNV analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.0.3 (2020-10-10)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: CentOS Linux 7 (Core)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /hpc/pmc_stunnenberg/cruiz/miniconda3/envs/r_pHGG_project/lib/libopenblasp-r0.3.12.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       "[1] infercnv_1.9.0\n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] nlme_3.1-152                bitops_1.0-7               \n",
       "  [3] matrixStats_0.59.0          doParallel_1.0.16          \n",
       "  [5] RColorBrewer_1.1-2          GenomeInfoDb_1.26.7        \n",
       "  [7] repr_1.1.3                  tools_4.0.3                \n",
       "  [9] utf8_1.2.2                  R6_2.5.0                   \n",
       " [11] KernSmooth_2.23-20          DBI_1.1.1                  \n",
       " [13] BiocGenerics_0.36.1         colorspace_2.0-2           \n",
       " [15] tidyselect_1.1.1            gridExtra_2.3              \n",
       " [17] compiler_4.0.3              argparse_2.0.3             \n",
       " [19] Biobase_2.50.0              formatR_1.11               \n",
       " [21] DelayedArray_0.16.3         sandwich_3.0-1             \n",
       " [23] caTools_1.18.2              scales_1.1.1               \n",
       " [25] mvtnorm_1.1-2               pbdZMQ_0.3-5               \n",
       " [27] digest_0.6.27               XVector_0.30.0             \n",
       " [29] base64enc_0.1-3             pkgconfig_2.0.3            \n",
       " [31] htmltools_0.5.1.1           parallelly_1.27.0          \n",
       " [33] MatrixGenerics_1.2.1        limma_3.46.0               \n",
       " [35] rlang_0.4.11                generics_0.1.0             \n",
       " [37] zoo_1.8-9                   jsonlite_1.7.2             \n",
       " [39] gtools_3.9.2                dplyr_1.0.7                \n",
       " [41] RCurl_1.98-1.3              magrittr_2.0.1             \n",
       " [43] modeltools_0.2-23           GenomeInfoDbData_1.2.4     \n",
       " [45] futile.logger_1.4.3         Matrix_1.3-3               \n",
       " [47] Rcpp_1.0.7                  IRkernel_1.1.1             \n",
       " [49] munsell_0.5.0               S4Vectors_0.28.1           \n",
       " [51] fansi_0.5.0                 reticulate_1.20            \n",
       " [53] ape_5.5                     lifecycle_1.0.0            \n",
       " [55] multcomp_1.4-17             edgeR_3.32.1               \n",
       " [57] MASS_7.3-54                 SummarizedExperiment_1.20.0\n",
       " [59] zlibbioc_1.36.0             plyr_1.8.6                 \n",
       " [61] gplots_3.1.1                grid_4.0.3                 \n",
       " [63] parallel_4.0.3              listenv_0.8.0              \n",
       " [65] crayon_1.4.1                lattice_0.20-44            \n",
       " [67] IRdisplay_1.0               splines_4.0.3              \n",
       " [69] locfit_1.5-9.4              pillar_1.6.1               \n",
       " [71] igraph_1.2.6                fastcluster_1.2.3          \n",
       " [73] GenomicRanges_1.42.0        uuid_0.1-4                 \n",
       " [75] codetools_0.2-18            stats4_4.0.3               \n",
       " [77] leiden_0.3.8                futile.options_1.0.1       \n",
       " [79] glue_1.4.2                  evaluate_0.14              \n",
       " [81] lambda.r_1.2.4              phyclust_0.1-30            \n",
       " [83] png_0.1-7                   vctrs_0.3.8                \n",
       " [85] foreach_1.5.1               tidyr_1.1.3                \n",
       " [87] gtable_0.3.0                RANN_2.6.1                 \n",
       " [89] purrr_0.3.4.9000            reshape_0.8.8              \n",
       " [91] future_1.21.0               assertthat_0.2.1           \n",
       " [93] ggplot2_3.3.5               coin_1.4-1                 \n",
       " [95] libcoin_1.0-8               coda_0.19-4                \n",
       " [97] rjags_4-10                  survival_3.2-11            \n",
       " [99] SingleCellExperiment_1.12.0 tibble_3.1.3               \n",
       "[101] iterators_1.0.13            IRanges_2.24.1             \n",
       "[103] globals_0.14.0              fitdistrplus_1.1-5         \n",
       "[105] TH.data_1.0-10              ellipsis_0.3.2             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(infercnv)\n",
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2021-07-25 09:48:02] Parsing matrix: data/gbm-count-matrix.tsv\n",
      "INFO [2021-07-25 09:55:55] Parsing gene order file: /hpc/pmc_stunnenberg/cruiz/scRNA/markers-and-databases/gencode_v38_gene_pos.txt\n",
      "INFO [2021-07-25 09:55:55] Parsing cell annotations file: data/gbm-anno-file.tsv\n",
      "INFO [2021-07-25 09:55:55] ::order_reduce:Start.\n",
      "INFO [2021-07-25 09:56:00] .order_reduce(): expr and order match.\n",
      "INFO [2021-07-25 09:56:05] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 25387,38830 Total=227223243 Min=0 Max=4059.\n",
      "INFO [2021-07-25 09:56:10] num genes removed taking into account provided gene ordering list: 823 = 3.24181667782723% removed.\n",
      "INFO [2021-07-25 09:56:10] -filtering out cells < 100 or > Inf, removing 0 % of cells\n",
      "INFO [2021-07-25 09:56:46] validating infercnv_obj\n"
     ]
    }
   ],
   "source": [
    "# create the infercnv object\n",
    "infercnv_obj = CreateInfercnvObject(raw_counts_matrix='data/gbm-count-matrix.tsv',\n",
    "                                    annotations_file='data/gbm-anno-file.tsv',\n",
    "                                    delim=\"\\t\",\n",
    "                                    gene_order_file=\"data/gencode_v38_gene_pos.txt\",\n",
    "                                    ref_group_names=c('CD4/CD8', 'DC', 'Endothelial', 'Mono',\n",
    "                                                     'Oligodendrocyte','Pericyte', \n",
    "                                                     'TAM-BDM', 'TAM-MG')\n",
    "                                    # Did not include predicted OPC, neuron and astrocytes to test whether\n",
    "                                    # the assigment of normal cells was correct \n",
    "                                    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2021-07-25 09:56:46] ::process_data:Start\n",
      "INFO [2021-07-25 09:56:46] Checking for saved results.\n",
      "INFO [2021-07-25 09:56:46] Trying to reload from step 17\n",
      "INFO [2021-07-25 09:57:01] Using backup HMM from step 17\n",
      "INFO [2021-07-25 09:57:01] Trying to reload from step 15\n",
      "INFO [2021-07-25 09:57:28] Using backup from step 15\n",
      "INFO [2021-07-25 09:57:28] \n",
      "\n",
      "\tSTEP 1: incoming data\n",
      "\n",
      "INFO [2021-07-25 09:57:28] \n",
      "\n",
      "\tSTEP 02: Removing lowly expressed genes\n",
      "\n",
      "INFO [2021-07-25 09:57:28] \n",
      "\n",
      "\tSTEP 03: normalization by sequencing depth\n",
      "\n",
      "INFO [2021-07-25 09:57:28] \n",
      "\n",
      "\tSTEP 04: log transformation of data\n",
      "\n",
      "INFO [2021-07-25 09:57:28] \n",
      "\n",
      "\tSTEP 08: removing average of reference data (before smoothing)\n",
      "\n",
      "INFO [2021-07-25 09:57:28] \n",
      "\n",
      "\tSTEP 09: apply max centered expression threshold: 3\n",
      "\n",
      "INFO [2021-07-25 09:57:28] \n",
      "\n",
      "\tSTEP 10: Smoothing data per cell by chromosome\n",
      "\n",
      "INFO [2021-07-25 09:57:28] \n",
      "\n",
      "\tSTEP 11: re-centering data across chromosome after smoothing\n",
      "\n",
      "INFO [2021-07-25 09:57:28] \n",
      "\n",
      "\tSTEP 12: removing average of reference data (after smoothing)\n",
      "\n",
      "INFO [2021-07-25 09:57:28] \n",
      "\n",
      "\tSTEP 14: invert log2(FC) to FC\n",
      "\n",
      "INFO [2021-07-25 09:57:28] \n",
      "\n",
      "\tSTEP 15: Clustering samples (not defining tumor subclusters)\n",
      "\n",
      "INFO [2021-07-25 09:57:47] \n",
      "\n",
      "\tSTEP 18: Run Bayesian Network Model on HMM predicted CNVs\n",
      "\n",
      "INFO [2021-07-25 09:57:49] Initializing new MCM InferCNV Object.\n",
      "INFO [2021-07-25 09:57:49] validating infercnv_obj\n",
      "INFO [2021-07-25 09:57:49] Total CNV's:  31\n",
      "INFO [2021-07-25 09:57:49] Loading BUGS Model.\n",
      "INFO [2021-07-25 09:57:50] Running Sampling Using Parallel with  8 Cores\n",
      "INFO [2021-07-25 18:35:45] Obtaining probabilities post-sampling\n",
      "INFO [2021-07-25 19:12:49] Gibbs sampling time:  554.995412977536  Minutes\n",
      "INFO [2021-07-25 19:17:17] ::plot_cnv:Start\n",
      "INFO [2021-07-25 19:17:17] ::plot_cnv:Current data dimensions (r,c)=8538,38830 Total=18616715.627469 Min=0 Max=0.794817954615511.\n",
      "INFO [2021-07-25 19:17:21] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2021-07-25 19:17:38] plot_cnv_observation:Start\n",
      "INFO [2021-07-25 19:17:38] Observation data size: Cells= 17529 Genes= 8538\n",
      "INFO [2021-07-25 19:17:44] plot_cnv_observation:Writing observations by grouping.\n",
      "INFO [2021-07-25 19:20:16] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2021-07-25 19:20:16] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2021-07-25 19:20:17] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2021-07-25 19:20:17] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2021-07-25 19:21:09] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2021-07-25 19:21:09] Quantiles of plotted data range: 0,0,0,0,0.794817954615511\n",
      "INFO [2021-07-25 19:21:47] plot_cnv_observations:Writing observation data to infercnv_FINAL/BayesNetOutput.HMMi6.hmm_mode-samples/infercnv.NormalProbabilities.PreFiltering.observations.txt\n",
      "INFO [2021-07-25 19:24:20] plot_cnv_references:Start\n",
      "INFO [2021-07-25 19:24:20] Reference data size: Cells= 21301 Genes= 8538\n",
      "INFO [2021-07-26 01:37:48] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2021-07-26 01:37:52] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2021-07-26 01:38:56] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2021-07-26 01:38:56] Quantiles of plotted data range: 0,0,0,0,0.638675984266198\n",
      "INFO [2021-07-26 01:39:23] plot_cnv_references:Writing reference data to infercnv_FINAL/BayesNetOutput.HMMi6.hmm_mode-samples/infercnv.NormalProbabilities.PreFiltering.references.txt\n",
      "INFO [2021-07-26 01:45:39] \n",
      "\n",
      "\tSTEP 19: Filter HMM predicted CNVs based on the Bayesian Network Model results and BayesMaxPNormal\n",
      "\n",
      "INFO [2021-07-26 01:45:39] Attempting to removing CNV(s) with a probability of being normal above  0.5\n",
      "INFO [2021-07-26 01:45:39] Removing  8  CNV(s) identified by the HMM.\n",
      "INFO [2021-07-26 01:45:42] Total CNV's after removing:  23\n",
      "INFO [2021-07-26 01:45:42] Reassigning CNVs based on state probabilities.\n",
      "INFO [2021-07-26 01:45:46] Changing the following CNV's states assigned by the HMM to the following based on the CNV's state probabilities.\n",
      " 1-region_2 : 4  (P= 0.280216554289716 ) ->  3 (P= 0.459132434048449 )\n",
      "1-region_4 : 4  (P= 0.28919027797633 ) ->  3 (P= 0.477806888199011 )\n",
      "3-region_12 : 4  (P= 0.286981518391902 ) ->  3 (P= 0.418067790137652 )\n",
      "4-region_15 : 4  (P= 0.322788550768369 ) ->  3 (P= 0.424094932486085 )\n",
      "5-region_16 : 4  (P= 0.317253179854884 ) ->  3 (P= 0.479996152031064 )\n",
      "6-region_19 : 4  (P= 0.346239438683596 ) ->  3 (P= 0.495832037219821 )\n",
      "7-region_22 : 4  (P= 0.283925871391475 ) ->  3 (P= 0.329875587568334 )\n",
      "7-region_26 : 4  (P= 0.308799179245656 ) ->  5 (P= 0.313551579834575 )\n",
      "9-region_34 : 4  (P= 0.265521436973709 ) ->  3 (P= 0.40328156010195 )\n",
      "14-region_45 : 2  (P= 0.43154656390259 ) ->  3 (P= 0.463574229067361 )\n",
      "6-region_200 : 4  (P= 0.0935303282045713 ) ->  3 (P= 0.450873584492748 )\n",
      "6-region_202 : 4  (P= 0.180988979402785 ) ->  3 (P= 0.362785922754425 )\n",
      "15-region_212 : 4  (P= 0.182436419381196 ) ->  3 (P= 0.361324015733802 )\n",
      "INFO [2021-07-26 01:45:46] Creating Plots for CNV and cell Probabilities.\n",
      "INFO [2021-07-26 01:55:37] ::plot_cnv:Start\n",
      "INFO [2021-07-26 01:55:37] ::plot_cnv:Current data dimensions (r,c)=8538,38830 Total=18388115.4637378 Min=0 Max=0.794817954615511.\n",
      "INFO [2021-07-26 01:55:41] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2021-07-26 01:55:56] plot_cnv_observation:Start\n",
      "INFO [2021-07-26 01:55:56] Observation data size: Cells= 17529 Genes= 8538\n",
      "INFO [2021-07-26 01:56:01] plot_cnv_observation:Writing observations by grouping.\n",
      "INFO [2021-07-26 01:58:33] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2021-07-26 01:58:33] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2021-07-26 01:58:34] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2021-07-26 01:58:34] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2021-07-26 01:59:19] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2021-07-26 01:59:19] Quantiles of plotted data range: 0,0,0,0,0.794817954615511\n",
      "INFO [2021-07-26 01:59:55] plot_cnv_observations:Writing observation data to infercnv_FINAL/BayesNetOutput.HMMi6.hmm_mode-samples/infercnv.NormalProbabilities.PostFiltering.observations.txt\n",
      "INFO [2021-07-26 02:02:37] plot_cnv_references:Start\n",
      "INFO [2021-07-26 02:02:37] Reference data size: Cells= 21301 Genes= 8538\n",
      "INFO [2021-07-26 08:16:18] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2021-07-26 08:16:22] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2021-07-26 08:17:22] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2021-07-26 08:17:22] Quantiles of plotted data range: 0,0,0,0,0.638675984266198\n",
      "INFO [2021-07-26 08:17:45] plot_cnv_references:Writing reference data to infercnv_FINAL/BayesNetOutput.HMMi6.hmm_mode-samples/infercnv.NormalProbabilities.PostFiltering.references.txt\n",
      "INFO [2021-07-26 08:22:06] ::plot_cnv:Start\n",
      "INFO [2021-07-26 08:22:06] ::plot_cnv:Current data dimensions (r,c)=8538,38830 Total=1005690802 Min=2 Max=5.\n",
      "INFO [2021-07-26 08:22:10] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2021-07-26 08:28:36] plot_cnv_observation:Start\n",
      "INFO [2021-07-26 08:28:36] Observation data size: Cells= 17529 Genes= 8538\n",
      "INFO [2021-07-26 08:28:41] plot_cnv_observation:Writing observations by grouping.\n",
      "INFO [2021-07-26 08:31:32] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2021-07-26 08:31:32] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2021-07-26 08:31:33] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2021-07-26 08:31:33] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2021-07-26 08:32:19] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2021-07-26 08:32:19] Quantiles of plotted data range: 2,3,3,3,5\n",
      "INFO [2021-07-26 08:32:55] plot_cnv_observations:Writing observation data to infercnv_FINAL/infercnv.19_HMM_pred.Bayes_Net.Pnorm_0.5.observations.txt\n",
      "INFO [2021-07-26 08:35:48] plot_cnv_references:Start\n",
      "INFO [2021-07-26 08:35:48] Reference data size: Cells= 21301 Genes= 8538\n",
      "INFO [2021-07-26 14:49:44] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2021-07-26 14:49:48] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2021-07-26 14:50:47] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2021-07-26 14:50:47] Quantiles of plotted data range: 3,3,3,3,3\n",
      "INFO [2021-07-26 14:51:10] plot_cnv_references:Writing reference data to infercnv_FINAL/infercnv.19_HMM_pred.Bayes_Net.Pnorm_0.5.references.txt\n",
      "INFO [2021-07-26 14:54:44] \n",
      "\n",
      "\tSTEP 20: Converting HMM-based CNV states to repr expr vals\n",
      "\n",
      "INFO [2021-07-26 14:56:30] ::plot_cnv:Start\n",
      "INFO [2021-07-26 14:56:30] ::plot_cnv:Current data dimensions (r,c)=8538,38830 Total=337080131 Min=0.5 Max=2.\n",
      "INFO [2021-07-26 14:56:35] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2021-07-26 15:03:22] plot_cnv_observation:Start\n",
      "INFO [2021-07-26 15:03:22] Observation data size: Cells= 17529 Genes= 8538\n",
      "INFO [2021-07-26 15:03:28] plot_cnv_observation:Writing observations by grouping.\n",
      "INFO [2021-07-26 15:06:26] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2021-07-26 15:06:28] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2021-07-26 15:06:29] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2021-07-26 15:06:30] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2021-07-26 15:07:17] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2021-07-26 15:07:17] Quantiles of plotted data range: 0.5,1,1,1,2\n",
      "INFO [2021-07-26 15:07:52] plot_cnv_observations:Writing observation data to infercnv_FINAL/infercnv.20_HMM_predHMMi6.hmm_mode-samples.Pnorm_0.5.repr_intensities.observations.txt\n",
      "INFO [2021-07-26 15:10:50] plot_cnv_references:Start\n",
      "INFO [2021-07-26 15:10:50] Reference data size: Cells= 21301 Genes= 8538\n",
      "INFO [2021-07-26 21:24:19] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2021-07-26 21:24:22] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2021-07-26 21:25:22] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2021-07-26 21:25:22] Quantiles of plotted data range: 1,1,1,1,1\n",
      "INFO [2021-07-26 21:25:45] plot_cnv_references:Writing reference data to infercnv_FINAL/infercnv.20_HMM_predHMMi6.hmm_mode-samples.Pnorm_0.5.repr_intensities.references.txt\n",
      "INFO [2021-07-26 21:29:23] \n",
      "\n",
      "\tSTEP 22: Denoising\n",
      "\n",
      "INFO [2021-07-26 21:29:23] ::process_data:Remove noise, noise threshold defined via ref mean sd_amplifier:  1.5\n",
      "INFO [2021-07-26 21:29:23] denoising using mean(normal) +- sd_amplifier * sd(normal) per gene per cell across all data\n",
      "INFO [2021-07-26 21:29:35] :: **** clear_noise_via_ref_quantiles **** : removing noise between bounds:  0.966477701403041 - 1.0344961111886\n",
      "INFO [2021-07-26 21:33:02] \n",
      "\n",
      "## Making the final infercnv heatmap ##\n",
      "INFO [2021-07-26 21:33:02] ::plot_cnv:Start\n",
      "INFO [2021-07-26 21:33:02] ::plot_cnv:Current data dimensions (r,c)=8538,38830 Total=332617714.785693 Min=0.824132073829285 Max=7.0561877854972.\n",
      "INFO [2021-07-26 21:33:07] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2021-07-26 21:40:30] plot_cnv(): auto thresholding at: (0.882854 , 1.117146)\n",
      "INFO [2021-07-26 21:40:47] plot_cnv_observation:Start\n",
      "INFO [2021-07-26 21:40:47] Observation data size: Cells= 17529 Genes= 8538\n",
      "INFO [2021-07-26 21:40:52] plot_cnv_observation:Writing observations by grouping.\n",
      "INFO [2021-07-26 21:44:06] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2021-07-26 21:44:07] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2021-07-26 21:44:08] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2021-07-26 21:44:08] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2021-07-26 21:45:02] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2021-07-26 21:45:02] Quantiles of plotted data range: 0.882854387706594,1.00048690629582,1.00048690629582,1.00048690629582,1.11714561229341\n",
      "INFO [2021-07-26 21:45:38] plot_cnv_observations:Writing observation data to infercnv_FINAL/infercnv.observations.txt\n",
      "INFO [2021-07-26 21:48:53] plot_cnv_references:Start\n",
      "INFO [2021-07-26 21:48:53] Reference data size: Cells= 21301 Genes= 8538\n",
      "INFO [2021-07-27 04:02:43] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2021-07-27 04:02:47] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2021-07-27 04:03:42] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2021-07-27 04:03:42] Quantiles of plotted data range: 0.882854387706594,1.00048690629582,1.00048690629582,1.00048690629582,1.11714561229341\n",
      "INFO [2021-07-27 04:04:06] plot_cnv_references:Writing reference data to infercnv_FINAL/infercnv.references.txt\n"
     ]
    }
   ],
   "source": [
    "# perform infercnv operations to reveal cnv signal\n",
    "infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                             cutoff=0.1, \n",
    "                             out_dir=\"infercnv_FINAL\", \n",
    "                             cluster_by_groups=F, \n",
    "                             denoise=TRUE,\n",
    "                             HMM=TRUE,\n",
    "                             num_threads = 8\n",
    "                             )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(infercnv_obj, 'data/Ruiz2021_infercnv.rds')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
